src/OmniSharp.Roslyn.CSharp/Helpers/LocationExtensions.cs

```diff
@@ -17,22 +19,55 @@ public static QuickFix GetQuickFix(this Location location, OmniSharpWorkspace wo
             var lineSpan = Path.GetExtension(location.SourceTree.FilePath).Equals(".cake", StringComparison.OrdinalIgnoreCase)
                 ? location.GetLineSpan()
                 : location.GetMappedLineSpan();
-            var path = lineSpan.Path;
-            var documents = workspace.GetDocuments(path);
 
-            var line = lineSpan.StartLinePosition.Line;
-            var text = location.SourceTree.GetText().Lines[line].ToString();
+            var documents = workspace.GetDocuments(lineSpan.Path);
+            var sourceText = GetSourceText(location, documents, lineSpan.HasMappedPath);
+            var text = GetText(location, sourceText, lineSpan.StartLinePosition.Line);
 
             return new QuickFix
             {
-                Text = text.Trim(),
-                FileName = path,
-                Line = line,
+                Text = text,
+                FileName = lineSpan.Path,
+                Line = lineSpan.StartLinePosition.Line,
                 Column = lineSpan.HasMappedPath ? 0 : lineSpan.StartLinePosition.Character, // when a #line directive maps into a separate file, assume columns (0,0)
                 EndLine = lineSpan.EndLinePosition.Line,
                 EndColumn = lineSpan.HasMappedPath ? 0 : lineSpan.EndLinePosition.Character,
                 Projects = documents.Select(document => document.Project.Name).ToArray()
             };
+
+            static SourceText GetSourceText(Location location, IEnumerable<Document> documents, bool hasMappedPath)
+            {
+                // if we have a mapped linespan and we found a corresponding document, pick that one
+                // otherwise use the SourceText of current location
+                if (hasMappedPath)
+                {
+                    if (documents != null && documents.Any() && documents.First().TryGetText(out SourceText sourceText))
+                    {
+                        // we have a mapped document that exists in workspace
+                        return sourceText;
+                    }
+
+                    // we have a mapped document that doesn't exist in workspace
+                    // in that case we have to always fall back to original linespan
+                    return null;
+                }
+
+                // unmapped document so just continue with current SourceText
+                return location.SourceTree.GetText();
+            }
+
+            static string GetText(Location location, SourceText sourceText, int startLine)
+            {
+                // bounds check in case the mapping is incorrect, since user can put whatever line they want
+                if (sourceText != null && sourceText.Lines.Count > startLine)
```

Comment:
Can SourceText actually be `null` here?

Can the user put a negative line mapping accidentally?

Reply:
1. yes if we can't find the mapped document in the workspace
2. yes - then the mapping doesn't resolve at in Roslyn and we will return the current file and current line number instead. I added an extra test for this so that the behavior is captured and won't change